### PR TITLE
Extending QuantileQuantizedType with quantileType mlir::Type member

### PR DESCRIPTION
**Pull request by sartil**
## Summary
WIP, as the functionality is being tested in the end to end enablement (also mlir tests might be expanded)

## JIRA ticket

* [E-137917](https://jira.devtools.intel.com/browse/EISW-137917)

## Related PR in NPU Compiler and/or OpenVINO repository with sub-module update

* [PR-12820](https://github.com/intel-innersource/applications.ai.vpu-accelerators.vpux-plugin/pull/12820)

* E-xxxxx

### CODE REVIEW

**Comment by sartil** [comment]
While using QuantileQuantizedType and QuantileQuantizedPerAxisType in the npu compiler for enabling palletization end to end, I noticed that both types can be upcast successfully to the base QuantizedType, but they can't be cast successfully instead to their parent types, UniformQuantizedType and UniformQuantizedPerAxisType respectively (dyn_cast unexpectedly fails returning nullptr).

This is a problem because it makes more expensive handling the two new types requiring an update in several locations to specifically handle them when instead both parent and child could be handled most of the times by just dynamic casting to the parent (in case only the quantization parameters are of interest in a function and not the quantile ones).

I tried to change the .td file in different ways as well as defining the `classof` method for the non-base classes (it's only defined for QuantTypes) but without success, the cast still doesn't succeed in all the cases.

@nikita-kud @andrey-golubev @ZoranZomborat when you have time could you have a quick look at QuantOpsBase.td and QuantTypes.h to see if you spot a problem?

**Review by ZoranZomborat — APPROVED** [review]
Interesting open you found related to casting ...

The changes themselves here look good!

**Comment by ZoranZomborat** [comment]
> While using QuantileQuantizedType and QuantileQuantizedPerAxisType in the npu compiler for enabling palletization end to end, I noticed that both types can be upcast successfully to the base QuantizedType, but they can't be cast successfully instead to their parent types, UniformQuantizedType and UniformQuantizedPerAxisType respectively (dyn_cast unexpectedly fails returning nullptr).
> 
> This is a problem because it makes more expensive handling the two new types requiring an update in several locations to specifically handle them when instead both parent and child could be handled most of the times by just dynamic casting to the parent (in case only the quantization parameters are of interest in a function and not the quantile ones).
> 
> I tried to change the .td file in different ways as well as defining the `classof` method for the non-base classes (it's only defined for QuantTypes) but without success, the cast still doesn't succeed in all the cases.
> 
> @nikita-kud @andrey-golubev @ZoranZomborat when you have time could you have a quick look at QuantOpsBase.td and QuantTypes.h to see if you spot a problem?

+ @hrotuna  also

**Comment by ZoranZomborat** [comment]
> While using QuantileQuantizedType and QuantileQuantizedPerAxisType in the npu compiler for enabling palletization end to end, I noticed that both types can be upcast successfully to the base QuantizedType, but they can't be cast successfully instead to their parent types, UniformQuantizedType and UniformQuantizedPerAxisType respectively (dyn_cast unexpectedly fails returning nullptr).
> 
> This is a problem because it makes more expensive handling the two new types requiring an update in several locations to specifically handle them when instead both parent and child could be handled most of the times by just dynamic casting to the parent (in case only the quantization parameters are of interest in a function and not the quantile ones).
> 
> I tried to change the .td file in different ways as well as defining the `classof` method for the non-base classes (it's only defined for QuantTypes) but without success, the cast still doesn't succeed in all the cases.
> 
> @nikita-kud @andrey-golubev @ZoranZomborat when you have time could you have a quick look at QuantOpsBase.td and QuantTypes.h to see if you spot a problem?

Possibly related to the LLVM deprecated casting method, ticket related https://jira.devtools.intel.com/browse/EISW-71627

**Comment by sartil** [comment]
The cast problem seems to be fixed now, just added a redefinition of `classof` function using type IDs.

**Comment by ZoranZomborat** [comment]
@nikita-kud @andrey-golubev @hrotuna please review also

**Review by ZoranZomborat — APPROVED** [review]

File: mlir/test/Dialect/Quant/Bytecode/types.mlir [hunk 1]
```diff
@@ -70,8 +70,8 @@ module @parseUniformPerAxisMixed attributes {
...
...00e-01,3.412000e-01,3.490000e-01,3.569000e-01,3.647000e-01,3.725000e-01,3.804000e-01,3.882000e-01,3.961000e-01,4.039000e-01,4.118000e-01,4.196000e-01,4.275000e-01,4.353000e-01,4.431000e-01,4.510000e-01,4.588000e-01,4.667000e-01,4.745000e-01,4.824000e-01,4.902000e-01,4.980000e-01,5.059000e-01,5.137000e-01,5.216000e-01,5.294000e-01,5.373000e-01,5.451000e-01,5.529000e-01,5.608000e-01,5.686000e-01,5.765000e-01,5.843000e-01,5.922000e-01,6.000000e-01,6.078000e-01,6.157000e-01,6.235000e-01,0.63139999999999996,0.63919999999999999,6.471000e-01,6.549000e-01,6.627000e-01,6.706000e-01,6.784000e-01,6.863000e-01,6.941000e-01,0.70199999999999996,7.098000e-01,7.176000e-01,7.255000e-01,7.333000e-01,7.412000e-01,7.490000e-01,7.569000e-01,7.647000e-01,7.725000e-01,0.78039999999999998,7.882000e-01,7.961000e-01,8.039000e-01,8.118000e-01,8.196000e-01,8.275000e-01,8.353000e-01,0.84309999999999996,8.510000e-01,8.588000e-01,8.667000e-01,8.745000e-01,8.824000e-01,8.902000e-01,8.980000e-01,9.059000e-01,0.91369999999999995,0.92159999999999997,9.294000e-01,9.373000e-01,9.451000e-01,9.529000e-01,0.96079999999999998,9.686000e-01,9.765000e-01,0.98429999999999995,0.99219999999999997,1.000000e+00}:9.987200e-01:127>
```
Comment (sartil) [inline mlir/test/Dialect/Quant/Bytecode/types.mlir:73]: this case is just for testing a generic type, but in our specific npu scenario, when palletization is enabled and quant.quantile types are in use the storage type is at most 4 bits wide, so the quantile LUT stored within the type has at most only 16 elements

**Review by hrotuna — APPROVED** [review]

File: mlir/include/mlir/Dialect/Quant/QuantTypes.h [hunk 2]
```diff
@@ -397,11 +401,12 @@ class UniformQuantizedPerAxisType
...
 /// Syntax synopsis:
 ///   Per-layer, all parameters expressed:
-///     !quant<quantile[StorageType:ExpressedType]{Quantiles}:{Scale:ZeroPoint}>
+///     !quant<quantile[StorageType:QuantileType:ExpressedType]{Quantiles}:{Scale:ZeroPoint}>
 ///   Per-layer, optional parameters omitted:
-///     !quant<quantile[StorageType]{Quantiles}:{Scale}>
+///     !quant<quantile[StorageType:QuantileType]{Quantiles}:{Scale}>
 ///
 ///   StorageType: 'i'|'u' NumBits
+///   QuantileType: 'i'|'u' NumBits, 'hf8', 'bf8', 'f16', 'bf16', 'f32', 'f64'
```
Comment (nikita-kud) [inline mlir/include/mlir/Dialect/Quant/QuantTypes.h:410]: It would be nice to add a little description of what this means.

**Review by nikita-kud — APPROVED** [review]
@sartil have you run `check-mlir` locally?

File: mlir/include/mlir/Dialect/Quant/QuantTypes.h (hunk 2 above)
Comment (sartil) [reply mlir/include/mlir/Dialect/Quant/QuantTypes.h:410]: sure, I will add it; I tried to replicate the pre-existing style for the description, which is a bit cryptic

Comment (sartil) [reply mlir/include/mlir/Dialect/Quant/QuantTypes.h:410]: Regarding `check-mlir`: I've run `mlir-opt` compiled with our build system and using FileCheck to check the mlir tests for the quant dialect. I'll run also `check-mlir`.

Comment (sartil) [reply mlir/include/mlir/Dialect/Quant/QuantTypes.h:410]: The report from check-mlir seems ok:

image

Comment (sartil) [reply mlir/include/mlir/Dialect/Quant/QuantTypes.h:410]: Added more verbose comment about what types are supported for quantileType member

**Comment by nikita-kud** [comment]
> @sartil have you run check-mlir locally?

There's a local run here: https://github.com/intel/npu-plugin-llvm/pull/60#discussion_r1785947842

**Pull request merged**